### ISSUE TITLE
Enhancement: [no-unnecessary-condition] check nullish switch cases

### ISSUE THREAD

**Issue by chenglou** [typescript-eslint/typescript-eslint#9802]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=MYewdgzgLgBAhgLhgcgILJgHxQIQwXhgCkBlAeQDkA6ABzgCcIBTACmSiemQEoAoXiAHcAllGAALGCzjcYAb14wYwOMxToEipcvAQQAGyZV9IAOZtxwnlqUAjekzgBrANxaVa5Hk3adkA0Ym5si2AJ5M1r72jq7uqkwwYACu%2BvpIAPTpMBDiICkAJjBM9PQg9DZ%2BeobGZmyC4nBQkdrRzrwAvvyZMMIAZlJwMPjD6jzyvN1KoP7VQRZWfN3tRfpqfQNDI8mpsnIw3cWljDA0pTTF%2BqETWVO6ATXBYRGLWe1AA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const a: 'A' | 'B' = JSON.parse('test')

switch (a) {
  case 'A':
    console.log('hi')
    break;
  case 'B':
    console.log('bye')
    break;
  case null: // should error
    console.log('what')
    break
}

// if (a === 'A') {
//   console.log('hi')
// } else if (a === null) { // errors properly
//   console.log('bye')
// }
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The null switch case should have errored

### Actual Result

No error

### Additional Info

Works fine (real error) with an if-else

**Comment by bradzacher** [comment]
The rule doesn't check switch cases at all as they're syntactically distinct from logical expressions and if conditions.

Technically they could be considered logical expressions though, yeah. It's not something anyone has ever considered.

It's not a bug as this was never a case that was considered for the rule. I'm happy to consider this feature request though. Seems like a good check to add.

**Comment by chenglou** [comment]
Oh yeah ok, then yes please! Many cases where we removed nullability from a type and didn't catch the dead code in other parts of the codebase that switch on the value
